### DISCUSSION TITLE
How do you get the FTP Session to End?

### DISCUSSION THREAD

**Discussion by jedhatchell** [pschatzmann/TinyFTPClient#5] [Q&A] [answered]
I've made good use of this client and really appreciate the effort that went into it, along with your timely response to my question the other day.  I've noticed something that I wanted to run by you and get your advice.

I'm using the "client.end()" method to end my sessions.  I was expecting this to end the ftp session to the server, similar to the ftp command "quit".  This is not working.  After the program reaches the end, even if I restart the program from scratch, the connection to the ftp server persists, and the file cannot be accessed because it is still in use.  Ultimately, the session time out, but a cleaner approach would be nice.

Is this the proper way to "quit" and ftp session.  If not, how?
Thanks again.
Joe

**Comment by pschatzmann** [comment, marked as answer]
end() is calling the server quit command. Did you check the result of your call (it should be true if it was successful) and if your server supports this command at all ?

You can also try to call stop() on the 2 client objects that you pass to force a disconnect

**Comment by jedhatchell** [comment]
I’m new to this so I don’t know yet how to read all the source code in your repository, so forgive me for what is perhaps an ignorant question.  I’m working primarily from the example code you’ve place there.  In all cases I see you end with this:

 

file.close();

client.end();

 

This is what I’m using.

 

Is this the end(), you’re referring to?  How would you “call stop()” on the 2 client objects?

 

Any help, very much appreciated.

**Comment by pschatzmann** [comment]
I recommend to check the result of your method calls.

So, if you have
```C++
    WiFiClient cmd;
    WiFiClient data;
    FTPClient client(cmd, data);
```
you can call

```C++
  // quit and display error if necessary
  if (!client.end()) {
      Serial.println("end() has failed");
      // force close connections
      data->stop();
      cmd->stop();
  }  
```
To check if your ftp server supports the quit command you login with a ftp client and enter quit: I am expecting 221 (=Goodbye).

**Comment by jedhatchell** [comment]
Got it.  Thanks again.
